lib/streamlit/state/session_state.py

```diff
@@ -54,27 +198,29 @@ class SessionState(MutableMapping[str, Any]):
     """
 
     _old_state: Dict[str, Any] = attr.Factory(dict)
-    _new_state: Dict[str, Any] = attr.Factory(dict)
-
-    def make_state_old(self) -> None:
-        self._old_state.update(self._new_state)
-        self._new_state.clear()
+    _new_session_state: Dict[str, Any] = attr.Factory(dict)
+    _new_widget_state: WStates = attr.Factory(WStates)
+
+    # TODO: widgets being moved to the old state dict should be unwrapped
+    # is it possible for a value to get through this without being deserialized?
+    def _compact_state(self) -> None:
+        self._old_state.update(self._new_session_state)
+        # self._old_state.update(self._new_widget_state)
+        for wid in self._new_widget_state:
+            self._old_state[wid] = self._new_widget_state[wid]
+        self._new_session_state.clear()
+        self._new_widget_state.clear()
 
     @property
     def _merged_state(self) -> Dict[str, Any]:
         # NOTE: The order that the dicts are unpacked here is important as it
-        #       is what ensures that the values in _new_state overwrite those
-        #       of _old_state in the returned, merged dictionary.
-        widget_mgr = _get_widget_mgr()
+        #       is what ensures that the new values take priority
         return {
             **self._old_state,
-            **widget_mgr.get_keyed_widget_values(),
-            **self._new_state,
+            **self._new_session_state,
```

Comment:
Should `_new_session_state` take priority over `_new_widget_state` here? In the case where a callback runs and writes to a `st.session_state` key corresponding to a widget, we'd want to use that value even if the widget value has been changed.

Reply:
Oh, I think that's right